Hi,
I tested with jenkins, works fine. For Jenkins it is enough to run tests with `tests.coverage=true` and configure the jenkins plugin to create the reports and save them inside jenkins. The task `jacocoTestReport` is obsolete for Jenkins.

I only had to tweak the settings a bit, because our build does not have default filenames (`*.exec`) or source location (`src/main/java`):

![image](https://user-images.githubusercontent.com/1005388/116781282-3533f500-aa82-11eb-9899-69486cf15343.png)

While running the jenkins Job, the replicator module produced a strange error with security manager:

```
> Task :lucene:replicator:test FAILED
Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:513)
	at java.instrument/sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:525)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jacoco.agent.rt.internal_f3994fa.core.runtime.InjectedClassRuntime$Lookup.privateLookupIn(InjectedClassRuntime.java:118)
	at org.jacoco.agent.rt.internal_f3994fa.core.runtime.InjectedClassRuntime.startup(InjectedClassRuntime.java:54)
	at org.jacoco.agent.rt.internal_f3994fa.PreMain.premain(PreMain.java:53)
	... 6 more
Caused by: java.security.AccessControlException: access denied ("java.lang.reflect.ReflectPermission" "suppressAccessChecks")
	at java.base/java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)
	at java.base/java.security.AccessController.checkPermission(AccessController.java:897)
	at java.base/java.lang.SecurityManager.checkPermission(SecurityManager.java:322)
	at java.base/java.lang.invoke.MethodHandles.privateLookupIn(MethodHandles.java:189)
	... 13 more
FATAL ERROR in native method: processing of -javaagent failed
```

So the replicator reports zero coverage.

Here are the results of first successful run on your PR's branch:
- Main build page: https://ci-builds.apache.org/job/Lucene/job/Lucene-Coverage-main/4/
- Go to the coverage reports on left ("Testabdeckung" in German, I got crazy to find this): https://ci-builds.apache.org/job/Lucene/job/Lucene-Coverage-main/4/jacoco/

--- a/gradle/testing/coverage.gradle
+++ b/gradle/testing/coverage.gradle
@@ -15,18 +15,49 @@
  * limitations under the License.
  */
 
-if (Boolean.parseBoolean(propertyOrDefault("tests.coverage", "false"))) {
+// This adds jacoco code coverage to tests.
+
+// Run with jacoco if either 'coverage' is passed as a task on input or
+// tests.coverage property is true.
+def withCoverage = gradle.startParameter.taskNames.contains("coverage") ||
+    Boolean.parseBoolean(propertyOrDefault("tests.coverage", "false"))
+
+if (withCoverage) {
   allprojects {
-    apply plugin: "jacoco"
-    tasks.withType(Test) {
-      // XXX: too many things called "workingDir" in gradle!
-      def targetDir = "${buildDir}/tmp/tests-cwd"
-      jacoco {
-        destinationFile = file("${targetDir}/jacoco.db")
+    plugins.withType(JavaPlugin) {
+      // Apply jacoco once we know the project has a Java plugin too.
+      project.plugins.apply("jacoco")
+
+      // Synthetic task to enable test coverage (and reports).
+      task coverage() {
+        dependsOn jacocoTestReport
       }
-      doFirst {
-        // delete any previous results
-        project.delete fileTree(dir: targetDir, include: 'jacoco.db')
+
+      // Deletes jacoco database prior to any tests.
+      task deleteCoverage(type: Delete) {
+      }
+
+      tasks.withType(Test) { Task testTask ->
+        // Configure jacoco destination file to be within the test
+        // task's working directory - this is related to security
+        // manager permissions (access to this file from within test jvm).
+        def jacocodb = file("${testTask.workingDir}/jacoco.exec")
+        jacoco {
+          destinationFile = jacocodb
+        }
+
+        // Register the database file for deletion. This ensures
+        // it's deleted once, before any test task runs.
+        deleteCoverage.delete jacocodb
+
+        // Test reports, if any, must be preceded by test execution.
+        jacocoTestReport.dependsOn testTask
+      }
+
+      configure(jacocoTestReport) {
+        doLast {
+          logger.lifecycle("Code coverage report at: ${reports.html.destination}.\n")
+        }
       }
     }
   }

--- a/gradle/testing/randomization/policies/replicator-tests.policy
+++ b/gradle/testing/randomization/policies/replicator-tests.policy
@@ -58,6 +58,7 @@ grant {
   permission java.lang.RuntimePermission "fileSystemProvider";
   // needed to test unmap hack on platforms that support it
   permission java.lang.RuntimePermission "accessClassInPackage.sun.misc";
+  permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
 };
 
 // Permissions to support ant build

--- a/gradle/testing/randomization/policies/replicator-tests.policy
+++ b/gradle/testing/randomization/policies/replicator-tests.policy
@@ -70,7 +70,7 @@ grant {
 // Permissions for jacoco code coverage
 grant {
   // permission to write the per-jvm code coverage output
-  permission java.io.FilePermission "${user.dir}${/}jacoco.db", "write";
+  permission java.io.FilePermission "${user.dir}${/}jacoco.exec", "write";
   // needed by jacoco to dump coverage on shutdown
   permission java.lang.RuntimePermission "shutdownHooks";
   // needed by jacoco to instrument classes

--- a/gradle/testing/randomization/policies/tests.policy
+++ b/gradle/testing/randomization/policies/tests.policy
@@ -109,7 +109,7 @@ grant {
 // Permissions for jacoco code coverage
 grant {
   // permission to write the per-jvm code coverage output
-  permission java.io.FilePermission "${user.dir}${/}jacoco.db", "write";
+  permission java.io.FilePermission "${user.dir}${/}jacoco.exec", "write";
   // needed by jacoco to dump coverage on shutdown
   permission java.lang.RuntimePermission "shutdownHooks";
   // needed by jacoco to instrument classes

--- a/help/tests.txt
+++ b/help/tests.txt
@@ -117,6 +117,7 @@ won't be the case (all tasks will use exactly the same starting seed):
 
 gradlew -p lucene/core beast -Ptests.dups=10 --tests TestPerFieldDocValuesFormat -Dtests.seed=deadbeef
 
+
 Verbose mode and debugging
 --------------------------
 
@@ -156,16 +157,18 @@ to increase the top-N count:
 
 gradlew -p lucene/core test -Ptests.profile=true -Ptests.profile.count=100
 
+
 Generating Coverage Reports
 ------------------------------
 
-Adding the property "tests.coverage=true" will run the tests with instrumentation to
-record coverage. The "jacocoTestReport" target will generate a report from those files.
+Running the "coverage" task (or setting the property "tests.coverage" to true)
+will run the tests with instrumentation to record code coverage.
 
 Example:
 
-gradlew -p lucene/core -Ptests.coverage=true test jacocoTestReport
+gradlew -p lucene/core coverage
 open lucene/core/build/reports/jacoco/test/html/index.html
+
 
 External data sets
 ------------------

--- a/lucene/replicator/src/test/org/apache/lucene/replicator/nrt/SimplePrimaryNode.java
+++ b/lucene/replicator/src/test/org/apache/lucene/replicator/nrt/SimplePrimaryNode.java
@@ -203,7 +203,7 @@ class SimplePrimaryNode extends PrimaryNode {
           c.out.writeByte(SimpleReplicaNode.CMD_PRE_COPY_MERGE);
           c.out.writeVLong(primaryGen);
           c.out.writeVInt(tcpPort);
-          SimpleServer.writeFilesMetaData(c.out, files);
+          TestSimpleServer.writeFilesMetaData(c.out, files);
           c.flush();
           c.s.shutdownOutput();
           message("warm connection " + c.s);
@@ -390,7 +390,7 @@ class SimplePrimaryNode extends PrimaryNode {
     out.writeBytes(state.infosBytes, 0, state.infosBytes.length);
     out.writeVLong(state.gen);
     out.writeVLong(state.version);
-    SimpleServer.writeFilesMetaData(out, state.files);
+    TestSimpleServer.writeFilesMetaData(out, state.files);
 
     out.writeVInt(state.completedMergeFiles.size());
     for (String fileName : state.completedMergeFiles) {
@@ -813,7 +813,7 @@ class SimplePrimaryNode extends PrimaryNode {
               c.out.writeByte(SimpleReplicaNode.CMD_PRE_COPY_MERGE);
               c.out.writeVLong(primaryGen);
               c.out.writeVInt(tcpPort);
-              SimpleServer.writeFilesMetaData(c.out, preCopy.files);
+              TestSimpleServer.writeFilesMetaData(c.out, preCopy.files);
               c.flush();
               c.s.shutdownOutput();
               message("successfully started warming");

--- a/lucene/replicator/src/test/org/apache/lucene/replicator/nrt/SimpleReplicaNode.java
+++ b/lucene/replicator/src/test/org/apache/lucene/replicator/nrt/SimpleReplicaNode.java
@@ -133,7 +133,7 @@ class SimpleReplicaNode extends ReplicaNode {
         // No incoming CopyState: ask primary for latest one now
         c.out.writeByte((byte) 1);
         c.flush();
-        copyState = SimpleServer.readCopyState(c.in);
+        copyState = TestSimpleServer.readCopyState(c.in);
         files = copyState.files;
       } else {
         c.out.writeByte((byte) 0);
@@ -331,7 +331,7 @@ class SimpleReplicaNode extends ReplicaNode {
 
           long newPrimaryGen = in.readVLong();
           curPrimaryTCPPort = in.readVInt();
-          Map<String, FileMetaData> files = SimpleServer.readFilesMetaData(in);
+          Map<String, FileMetaData> files = TestSimpleServer.readFilesMetaData(in);
           message("done reading files to copy files=" + files.keySet());
           AtomicBoolean finished = new AtomicBoolean();
           launchPreCopyMerge(finished, newPrimaryGen, files);

--- a/lucene/replicator/src/test/org/apache/lucene/replicator/nrt/TestNRTReplication.java
+++ b/lucene/replicator/src/test/org/apache/lucene/replicator/nrt/TestNRTReplication.java
@@ -39,7 +39,9 @@ import org.apache.lucene.util.LuceneTestCase;
 import org.apache.lucene.util.LuceneTestCase.SuppressCodecs;
 import org.apache.lucene.util.LuceneTestCase.SuppressSysoutChecks;
 import org.apache.lucene.util.SuppressForbidden;
+import org.apache.lucene.util.TestRuleIgnoreTestSuites;
 import org.apache.lucene.util.TestUtil;
+import org.apache.lucene.util.VerifyTestClassNamingConvention;
 
 // MockRandom's .sd file has no index header/footer:
 @SuppressCodecs({"MockRandom", "Direct", "SimpleText"})
@@ -93,17 +95,19 @@ public class TestNRTReplication extends LuceneTestCase {
       cmd.add("-Dtests.nrtreplication.forcePrimaryVersion=" + forcePrimaryVersion);
     }
 
+    // Mark as running nested.
+    cmd.add("-D" + TestRuleIgnoreTestSuites.PROPERTY_RUN_NESTED + "=true");
+
     // Mixin our own counter because this is called from a fresh thread which means the seed
     // otherwise isn't changing each time we spawn a
     // new node:
     long seed = random().nextLong() * nodeStartCounter.incrementAndGet();
-
     cmd.add("-Dtests.seed=" + SeedUtils.formatSeed(seed));
     cmd.add("-ea");
     cmd.add("-cp");
     cmd.add(System.getProperty("java.class.path"));
     cmd.add("org.junit.runner.JUnitCore");
-    cmd.add(getClass().getName().replace(getClass().getSimpleName(), "SimpleServer"));
+    cmd.add(TestSimpleServer.class.getName());
 
     message("child process command: " + cmd);
     ProcessBuilder pb = new ProcessBuilder(cmd);

--- a/lucene/replicator/src/test/org/apache/lucene/replicator/nrt/TestSimpleServer.java
+++ b/lucene/replicator/src/test/org/apache/lucene/replicator/nrt/TestSimpleServer.java
@@ -46,8 +46,11 @@ import org.apache.lucene.util.LuceneTestCase;
 import org.apache.lucene.util.LuceneTestCase.SuppressCodecs;
 import org.apache.lucene.util.LuceneTestCase.SuppressSysoutChecks;
 import org.apache.lucene.util.SuppressForbidden;
+import org.apache.lucene.util.TestRuleIgnoreTestSuites;
 import org.apache.lucene.util.TestUtil;
+import org.junit.Assume;
 import org.junit.AssumptionViolatedException;
+import org.junit.BeforeClass;
 
 /**
  * Child process with silly naive TCP socket server to handle between-node commands, launched for
@@ -56,7 +59,7 @@ import org.junit.AssumptionViolatedException;
 @SuppressCodecs({"MockRandom", "Direct", "SimpleText"})
 @SuppressSysoutChecks(bugUrl = "Stuff gets printed, important stuff for debugging a failure")
 @SuppressForbidden(reason = "We need Unsafe to actually crush :-)")
-public class SimpleServer extends LuceneTestCase {
+public class TestSimpleServer extends LuceneTestCase {
 
   static final Set<Thread> clientThreads = Collections.synchronizedSet(new HashSet<>());
   static final AtomicBoolean stop = new AtomicBoolean();
@@ -220,6 +223,11 @@ public class SimpleServer extends LuceneTestCase {
     long primaryGen = in.readVLong();
 
     return new CopyState(files, version, gen, infosBytes, completedMergeFiles, primaryGen, null);
+  }
+
+  @BeforeClass
+  public static void ensureNested() {
+    Assume.assumeTrue(TestRuleIgnoreTestSuites.isRunningNested());
   }
 
   @SuppressWarnings("try")

--- a/lucene/replicator/src/test/org/apache/lucene/replicator/nrt/TestStressNRTReplication.java
+++ b/lucene/replicator/src/test/org/apache/lucene/replicator/nrt/TestStressNRTReplication.java
@@ -51,6 +51,7 @@ import org.apache.lucene.util.LuceneTestCase;
 import org.apache.lucene.util.LuceneTestCase.SuppressCodecs;
 import org.apache.lucene.util.LuceneTestCase.SuppressSysoutChecks;
 import org.apache.lucene.util.SuppressForbidden;
+import org.apache.lucene.util.TestRuleIgnoreTestSuites;
 import org.apache.lucene.util.TestUtil;
 import org.apache.lucene.util.ThreadInterruptedException;
 
@@ -607,17 +608,19 @@ public class TestStressNRTReplication extends LuceneTestCase {
     long myPrimaryGen = primaryGen;
     cmd.add("-Dtests.nrtreplication.primaryGen=" + myPrimaryGen);
 
+    // Mark as running nested.
+    cmd.add("-D" + TestRuleIgnoreTestSuites.PROPERTY_RUN_NESTED + "=true");
+
     // Mixin our own counter because this is called from a fresh thread which means the seed
     // otherwise isn't changing each time we spawn a
     // new node:
     long seed = random().nextLong() * nodeStartCounter.incrementAndGet();
-
     cmd.add("-Dtests.seed=" + SeedUtils.formatSeed(seed));
     cmd.add("-ea");
     cmd.add("-cp");
     cmd.add(System.getProperty("java.class.path"));
     cmd.add("org.junit.runner.JUnitCore");
-    cmd.add(getClass().getName().replace(getClass().getSimpleName(), "SimpleServer"));
+    cmd.add(TestSimpleServer.class.getName());
 
     Writer childLog;
 

